### ISSUE TITLE
cudacpp plugin hardcode some change that modifies code outside of the sandbox of the plugin

### ISSUE THREAD

**Issue by oliviermattelaer** [madgraph5/madgraph4gpu#341]
Hi Andrea,

I was implementing the multi-channel for the standalone_gpu class, and was surprised that some object were pointing to your PLUGIN. looks like model_hangling.py is directly overwritting references to class in other file of mg5amc, which creates side effect to other codes.

Are those for debugging? if not then we should find a better way for you to specify the class that you want (via class attribute and subclassing)

Cheers,

Olivier

**Comment by valassi** [comment]
Hi Olivier,
no some things may have been for debugging, but some are not.

For instance this one was about disabling formatting (but ok, still aesthetics)
https://github.com/madgraph5/madgraph4gpu/blob/5ac285cebc04cd7eb4d6679c6a3a5e2645803d30/epochX/cudacpp/CODEGEN/PLUGIN/CUDACPP_SA_OUTPUT/model_handling.py#L47

We should probably sit down (even in zoom for the moment) and discuss the issues one by one (or start with the general ideas).

I agree, eventually the plugin stuff should get better integrated into the standard things, we need to discuss how to do that.

By the way one thing I can think of is that I would (and have not done yet) change very significantly CPPProcess, both in the class itself and in how it is generated. 

Anyway, its good that we are already at the level of converging and cleaning up, somehow :-)

You are right in assigning this to me, but for th emoment I cannot do much. I need to discuss with you the details first to understand. Or maybe I can start to have a look when/if you tell me that I should port my stuff to a specific more recent branch and commit of the code - when this is possible, let me know!

Andrea

PS This is also vaguely related to helicity recyling #279... more generally, I am well aware that my plugin works on top of a specific version of 270gpu, while it should be ported to more recent versions of the code (and be integrated in a more maintainable way).

**Comment by valassi** [comment]
To be discussed within #360 - upgrading from launchpad 270gpu to 311lovectorization

**Comment by valassi** [comment]
Hi @oliviermattelaer thanks for the discussion today.

Ok I understand better your concern now: the problem is that, if my plugin is present in a Madgraph installation, this presently affects other outputs even if my output is not requested at all. Clearly this isa problem.

I guess there are two ways to address this, both of which should be looked at

1. I can try to make the plugin more encapsulated. For instance, replace export_cpp/misc etc components only if my plugin is used, not all the time if it is present. 

While some of the hacks I implemented are quite radical in replacing the default code, I think that it is still quite useful to keep in mind this option that Python offers. I should just try to clean this up better.

2. The second complementary approach is to go through each and every change that I did outside the default plugin mechanism that you designed. For the moment we identified the following:

- (TODO) Replace # by // in export_cpp.get_mg5_info_lines
https://github.com/madgraph5/madgraph4gpu/blob/6f8c15a2bea4c4e2607c9bc377248ef9b666ca17/epochX/cudacpp/CODEGEN/PLUGIN/CUDACPP_SA_OUTPUT/model_handling.py#L25

- (DEBUG ONLY) add a debug printout in writers.FileWriter.__init__
https://github.com/madgraph5/madgraph4gpu/blob/6f8c15a2bea4c4e2607c9bc377248ef9b666ca17/epochX/cudacpp/CODEGEN/PLUGIN/CUDACPP_SA_OUTPUT/model_handling.py#L37

- (TODO) disable formatting in CPPWriter: note that I am doing the formattting in each method, I am NOT defining a new formatter that fixes issues... the new formatter I add is a totally dummy formatter that does not change anything
https://github.com/madgraph5/madgraph4gpu/blob/6f8c15a2bea4c4e2607c9bc377248ef9b666ca17/epochX/cudacpp/CODEGEN/PLUGIN/CUDACPP_SA_OUTPUT/model_handling.py#L47

- (?CHECK) disable caching? not clear if it is still needed, maybe I just used this for debugging
https://github.com/madgraph5/madgraph4gpu/blob/6f8c15a2bea4c4e2607c9bc377248ef9b666ca17/epochX/cudacpp/CODEGEN/PLUGIN/CUDACPP_SA_OUTPUT/model_handling.py#L62

- (DEBUG ONLY?) this is proably also just a debug printout
https://github.com/madgraph5/madgraph4gpu/blob/6f8c15a2bea4c4e2607c9bc377248ef9b666ca17/epochX/cudacpp/CODEGEN/PLUGIN/CUDACPP_SA_OUTPUT/model_handling.py#L73

- (TODO - EASY?) use a custom HelasWriter: you said that there should be an easier way to do this, through the plugin hooks that you already provide
https://github.com/madgraph5/madgraph4gpu/blob/6f8c15a2bea4c4e2607c9bc377248ef9b666ca17/epochX/cudacpp/CODEGEN/PLUGIN/CUDACPP_SA_OUTPUT/model_handling.py#L1525

- (TODO - NO LONGER NEEDED?) this was about handling sets and ordered dictionaries to make the output reproducible, but you fixed it upstream and in principle I can remove this?
https://github.com/madgraph5/madgraph4gpu/blob/6f8c15a2bea4c4e2607c9bc377248ef9b666ca17/epochX/cudacpp/CODEGEN/PLUGIN/CUDACPP_SA_OUTPUT/output.py#L36

- (TODO - NO LONGER NEEDED?) preserve symlinks in the 'cp' method: maybe however I do not need this anymore because I changed the way the plugin is defined? to nbe checked, else port upstream
https://github.com/madgraph5/madgraph4gpu/blob/6f8c15a2bea4c4e2607c9bc377248ef9b666ca17/epochX/cudacpp/CODEGEN/PLUGIN/CUDACPP_SA_OUTPUT/output.py#L55

Looking at the code, in principle this is all. I will go through each item.

**Comment by valassi** [comment]
Hi @oliviermattelaer thanks for the chat this morning!

About this specific issue, a couple of comments

1) I am now able to reproduce the issue when trying to generate madevent+standalone cpp/gpu (or probably also just standalone cpp/gpu), if the cudacpp plugin is in the plugin directory. This is because the plugin is imported even if it is not used later on. By simply importing it, some methods and classes in export_cpp are modified, in ways which break the standalone cpp and gpu backends from export_cpp.py.

This is obvious for instance in PR #401. This commit https://github.com/madgraph5/madgraph4gpu/pull/401/commits/d251848c1c7b9e26f60557ae7562fc3cb8389f84 is something that I had to add as a workaround. Essentially, I need to copy the cudacpp plugin ONLY when it is needed.

2) As discussed, I am actually able to produce madevent+cudacpp in PR #401. If I do
```
generate g g > t t~
output madevent CODEGEN_mad_gg_tt --hel_recycling=False --vector_size=16 --me_exporter=standalone_cudacpp
```
then this is able to create both fortran and cudacpp MEs. 

Well, I should check that the Fortran MEs are not affected.... This prprobably makes sense, as what I modify in the plugin is export_cpp which is used for cpp and gpu but not Fortran.

Anyway, this issue #341 afftects the gpu an dcpp outputs, NOT the new cudacpp output which is our main goal now. So, solving this #341 can be don elater (after #400, for instance).

3) As a temporary strategy to address #341 (the eventual final one being removing the export_cpp replavcements), I condidered another hack like
- define the new functions when you import the plugin
- replace export_cpp.function = new function ONLY in some init function for the plugin
- replace back export_cpp.function = the default export_cpp function in some exit function for the plugin

This probably is unfeasible as there are no such init/exit python mfunctions.

Anyway, Olivier suggest that output.PLUGIN_ProcessExporter is the main entry point to the plugin. So if there is any hope of finding entry/exit hooks, they would probably b ethere.

**Comment by valassi** [comment]
I had an idea that I will need to try out:
```
import export_cpp as export_cudacpp
```
If I do this in all plugin files, in principle this should build a parallel tree: so the original export_cpp used by cpp and gpu standalone remains unaffected, while export_cudacpp used by the cudacpp plugin is modified as much as I want. (I assume "import as" does effectively some sort of deepcopy... well otherwise one can deepcopy export_cpp as export_cudacpp, in a similar way). To be tested.

**Comment by valassi** [comment]
I still need to go one by one eventually.

I can remove misc.make_unique. This was just to remove a printout, now removed (was probably removed already in a previous branch, not in 311)
```
git show e811d84c0 madgraph/various/misc.py 
commit e811d84c04b339f664d999f895434835cf615632
Merge: beae73f2b c6443d97f
Author: Olivier Mattelaer <olivier.mattelaer@uclouvain.be>
Date:   Fri Jan 7 14:34:55 2022 +0100

    test of merge via git

diff --cc madgraph/various/misc.py
index 26f40a1fe,ec8b3399e..ed8c9baa5
--- a/madgraph/various/misc.py
+++ b/madgraph/various/misc.py
@@@ -2188,8 -2189,10 +2189,11 @@@ def make_unique(input, keepordering=Non
      "remove duplicate in a list "
  
      if keepordering is None:
-         keepordering = madgraph.ordering
-     sprint(keepordering)
```

**Comment by valassi** [comment]
Hm, some debugging, the issue seems to come from this redefinition
```
DEFAULT_GPUFOHelasCallWriter = helas_call_writers.GPUFOHelasCallWriter
helas_call_writers.GPUFOHelasCallWriter = PLUGIN_GPUFOHelasCallWriter
```

**Comment by valassi** [comment]
I have created a WIP PR #433

**Comment by valassi** [comment]
This is now finally fixed in PR #433. Closing.
